### PR TITLE
port message_filter from ROS to ROS2

### PR DESCRIPTION
**Pull request by jwang11**
The porting work include below updates:
*) change build tool from catkin to ament in CMakeLists.txt
*) ROS APIs move to ROS2 APIs
*) drop boost dependency, using C++14 instead
*) add message_event.h message_trait.h parameter_adapter.h which
   are missing in ROS2 core
*) existing test cases update and add a fuzz test(default disable)
*) drop unused setup.py
*) add a README

Signed-off-by: jwang <jing.j.wang@intel.com>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,90 +1,118 @@
...
-    target_link_libraries(${PROJECT_NAME}-test_subscriber message_filters)
+  ament_add_gtest(${PROJECT_NAME}-test_approximate_time_policy test/test_approximate_time_policy.cpp)
+  if(TARGET ${PROJECT_NAME}-test_approximate_time_policy)
+    target_link_libraries(${PROJECT_NAME}-test_approximate_time_policy ${PROJECT_NAME})
   endif()
 
-  # Unit test of the approximate synchronizer
-  catkin_add_nosetests(test/test_approxsync.py)
-  catkin_add_nosetests(test/test_message_filters_cache.py)
+#  ament_add_gtest(${PROJECT_NAME}-test_fuzz test/test_fuzz.cpp)
```
Comment (gaoethan) [inline CMakeLists.txt]: If the fuzz test defaults to disable, it's better deal with the  ROS2 way by `SKIP_TEST` with a macro to bypass it, then enabling it by give the macro defined to `--cmake-args` when necessary, and I can do this work later.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+[----------] 1 test from Subscriber
+[ RUN      ] Subscriber.fuzz_subscriber
+[       OK ] Subscriber.fuzz_subscriber (5010 ms)
+[----------] 1 test from Subscriber (5010 ms total)
+
+[----------] Global test environment tear-down
+[==========] 3 tests from 3 test cases ran. (15142 ms total)
+[  PASSED  ] 3 tests.
+```
+Any security issue should be reported using process at https://01.org/security."
```
Comment (gaoethan) [inline README.md]: Is it possible to add one bullet - `Known Issues` to allow the other to get aware of the legacy work to do or some issues that we'll address in the near future if we actually have. thanks !

File: include/message_filters/cache.h [hunk 3]
```diff
@@ -127,11 +128,12 @@ class Cache : public SimpleFilter<M>
    */
   void add(const EventType& evt)
   {
-    namespace mt = ros::message_traits;
+//    namespace mt = ros::message_traits;
```
Comment (gaoethan) [inline include/message_filters/cache.h]: It's better remove these codes unnecessary now, thanks !

File: include/message_filters/macros.h [hunk 4]
```diff
@@ -28,15 +28,15 @@
 #ifndef MESSAGE_FILTERS_MACROS_H_
 #define MESSAGE_FILTERS_MACROS_H_
 
-#include <ros/macros.h> // for the DECL's
-
+//#include <ros/macros.h> // for the DECL's
```
Comment (gaoethan) [inline include/message_filters/macros.h]: It's better remove these ROS codes unnecessary now, thanks !

File: include/message_filters/subscriber.h [hunk 5]
```diff
@@ -60,7 +58,8 @@ class SubscriberBase
    * \param transport_hints The transport hints to pass along
    * \param callback_queue The callback queue to pass along
    */
-  virtual void subscribe(ros::NodeHandle& nh, const std::string& topic, uint32_t queue_size, const ros::TransportHints& transport_hints = ros::TransportHints(), ros::CallbackQueueInterface* callback_queue = 0) = 0;
+//  virtual void subscribe(ros::NodeHandle& nh, const std::string& topic, uint32_t queue_size, const ros::TransportHints& transport_hints = ros::TransportHints(), ros::CallbackQueueInterface* callback_queue = 0) = 0;
```
Comment (gaoethan) [inline include/message_filters/subscriber.h]: It's better remove these ROS codes unnecessary now, thanks !

File: include/message_filters/subscriber.h [hunk 6]
```diff
@@ -109,9 +108,10 @@ class Subscriber : public SubscriberBase, public SimpleFilter<M>
    * \param transport_hints The transport hints to pass along
    * \param callback_queue The callback queue to pass along
    */
-  Subscriber(ros::NodeHandle& nh, const std::string& topic, uint32_t queue_size, const ros::TransportHints& transport_hints = ros::TransportHints(), ros::CallbackQueueInterface* callback_queue = 0)
+//  Subscriber(ros::NodeHandle& nh, const std::string& topic, uint32_t queue_size, const ros::TransportHints& transport_hints = ros::TransportHints(), ros::CallbackQueueInterface* callback_queue = 0)
```
Comment (gaoethan) [inline include/message_filters/subscriber.h]: It's better remove these ROS codes unnecessary now, thanks !

File: include/message_filters/subscriber.h [hunk 7]
```diff
@@ -137,16 +137,19 @@ class Subscriber : public SubscriberBase, public SimpleFilter<M>
    * \param transport_hints The transport hints to pass along
    * \param callback_queue The callback queue to pass along
    */
-  void subscribe(ros::NodeHandle& nh, const std::string& topic, uint32_t queue_size, const ros::TransportHints& transport_hints = ros::TransportHints(), ros::CallbackQueueInterface* callback_queue = 0)
+//  void subscribe(ros::NodeHandle& nh, const std::string& topic, uint32_t queue_size, const ros::TransportHints& transport_hints = ros::TransportHints(), ros::CallbackQueueInterface* callback_queue = 0)
```
Comment (gaoethan) [inline include/message_filters/subscriber.h]: It's better remove these ROS codes unnecessary now, thanks !

File: include/message_filters/subscriber.h [hunk 8]
```diff
@@ -169,18 +172,19 @@ class Subscriber : public SubscriberBase, public SimpleFilter<M>
...
   {
-    return ops_.topic;
+    return this->topic_;
   }
 
   /**
    * \brief Returns the internal ros::Subscriber object
    */
-  const ros::Subscriber& getSubscriber() const { return sub_; }
+//  const ros::Subscriber& getSubscriber() const { return sub_; }
```
Comment (gaoethan) [inline include/message_filters/subscriber.h]: Unnecessary ROS and better remove it. thanks !

File: test/msg_cache_unittest.cpp [hunk 9]
```diff
@@ -218,8 +218,9 @@ TEST(Cache, eventInEventOut)
 
 int main(int argc, char **argv){
   testing::InitGoogleTest(&argc, argv);
-  ros::init(argc, argv, "blah");
-  ros::Time::init();
+//  ros::init(argc, argv, "blah");
```
Comment (gaoethan) [inline test/msg_cache_unittest.cpp]: Unnecessary ROS and better remove it. thanks !

File: test/test_approxsync.py [hunk 10]
```diff
@@ -114,14 +116,13 @@ def test_approx(self):
             self.assertEqual(self.collector, [])
             for i in random.sample(range(N), N):
                 msg = seq1[i]
-                rospy.rostime._set_rostime(rospy.Time(i+0.05))
+#               rospy.rostime._set_rostime(rospy.Time(i+0.05))
```
Comment (gaoethan) [inline test/test_approxsync.py]: If it's unnecessary ROS, it's better remove it. thanks !

File: include/message_filters/null_types.h [hunk 11]
```diff
@@ -37,19 +37,28 @@
 
 #include "connection.h"
 
-#include <boost/shared_ptr.hpp>
-#include <ros/time.h>
-
-#include <ros/message_traits.h>
+//#include <ros/time.h>
```
Comment (gaoethan) [inline include/message_filters/null_types.h]: remove unnecessary ROS code, thanks !

**Comment by gaoethan** [comment]
Thanks a lot for Jing's contribution, I think we'd better add a bullet `known issues` in the readme if we actually have and clean those unnecessary ROS code. :)

**Comment by gaoethan** [comment]
lgtm now. if you approves these initial porting @testkit, I'll merge it, thanks !

File: README.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
```
Comment (testkit) [inline README.md]: Pls. remove the ' " ' at the end of the last line.

File: README.md (hunk 2 above)
Comment (testkit) [reply README.md]: @jwang11 , pls. remove the ' " ' at the end of last line in README.

File: README.md (hunk 12 above)
Comment (gaoethan) [reply README.md]: @testkit Cathy, Do you have any other review comments or concerns for Jing's initial porting except for the previous, Maybe Jing prefers not to fix these minor style issues and I'll optimize these style issues after merging this PR if no further concern in order to not pend it here for a long time, thanks !

**Pull request merged**